extensions/chathistory.md

```diff
@@ -0,0 +1,115 @@
+
+The server as mentioned in this document may refer to either an IRC server or an IRC bouncer.
+
+## Implementation
+The `chathistory` extension uses the [chathistory][batch/chathistory] batch type and introduces a client command, `chathistory`.
+
+To fully support this extension, clients MUST support the [`batch`][batch], [`server-time`][server-time] and [`message-tags`][message-tags] capabilities.
+
+The `chathistory` capability MUST be negotiated. This allows the server and client to act differently when delivering message history on connection.
+ 
+An ISUPPORT token MUST be sent to the client to state the maximum number of messages a client can request in a single command, represented by an integer. `CHATHISTORY=50`. If `0`, the client SHOULD assume that there is no maximum number of messages.
+
+### `CHATHISTORY` Command
+`CHATHISTORY` content can be requested by the client by sending the `CHATHISTORY` command to the server. A `batch` MUST be returned by the server. If no content exists to return, an empty batch SHOULD be returned to avoid the client waiting for a reply and to indicate that no content is available.
+
+The `chathistory` command uses the following general syntax structure:
+
+    CHATHISTORY <subcommand> <target> <timestamp | msgid> <limit>
+
+The `target` parameter specifies a single buffer (channel or nickname) from which history SHOULD be retrieved. Any `timestamp` values or parameters MUST be in the format of `YYYY-MM-DDThh:mm:ss.sssZ`.
+
+If a nickname is given as the `target` then the server SHOULD include history sent between both the current user and the target nickname as to give a full conversation. The server SHOULD attempt to include history involving other nicknames if either the current user or the target nickname has changed during the requested timeframe.
+
+#### Subcommands
+
+The following subcommands are used to describe how the server should return messages relative to the `timestamp` or `msgid` given.
+
+#### `BEFORE`
+
+    CHATHISTORY BEFORE <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+    
+Request up to `limit` number of messages before and including the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `AFTER`
+    CHATHISTORY AFTER <target> <timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request up to `limit` number of messages after and excluding the given `timestamp` or `msgid`. Only one timestamp or msgid MUST be given, not both.
+
+#### `LATEST`
+    CHATHISTORY LATEST <target> <* | timestamp=YYYY-MM-DDThh:mm:ss.sssZ | msgid=1234> <limit>
+Request the most recent messages that have been sent after and excluding the given `timestamp` or `msgid`. If a `*` is given instead of a timestamp or msgid, the server MUST use the current time as a timestamp. The number of messages returned MUST be equal to or less than `limit`. If a `*` is not given, only one timestamp or msgid MUST be given, not both.
```

Comment:
Multiple messages might have the exact same server-time; as a client I'd likely want to provide an inclusive timestamp instead of exclusive to the result and do de-duplication based on msgid locally. Otherwise there can be cases where I disconnected while receiving a series of messages at the exact same time and loose one of them.

It might even be better to ask for an older timestamp from a few minutes before and handle msgid based deduplication. If there are messages I didn't receive but have an older timestamp due to originating from a different linked server which can have latency between the server I am connected.

msgid lookup can face similiar problems.